### PR TITLE
add user agent and use case info to faucet json body

### PR DESCRIPTION
**Pull request by jonathanlei**
## High Level Overview of Change

aadd user agent and use case info to faucet transactions for analytics

### Context of Change

This is part of the testnet analytics effort.

### Type of Change
[ X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 1]
```diff
@@ -135,7 +135,10 @@ async def _check_wallet_balance(address: str, client: Client) -> int:
 
 async def _request_funding(url: str, address: str) -> None:
```
Comment (ckniffen) [inline xrpl/asyncio/wallet/wallet_generation.py:136]: I think an optional parameter of usecase would be helpful.  That way we can say 'unit test' or 'integration test'.

Comment (jonathanlei) [reply xrpl/asyncio/wallet/wallet_generation.py]: added

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 2]
```diff
@@ -31,6 +31,7 @@ async def generate_faucet_wallet(
     wallet: Optional[Wallet] = None,
     debug: bool = False,
     faucet_host: Optional[str] = None,
+    use_case: Optional[str] = None,
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:34]: Why not just call this param `user_agent`? `use_case` is a very confusing name IMO

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 3]
```diff
@@ -133,9 +138,18 @@ async def _check_wallet_balance(address: str, client: Client) -> int:
...
 
 
-async def _request_funding(url: str, address: str) -> None:
+async def _request_funding(
+    url: str, address: str, use_case: Optional[str] = None
+) -> None:
     async with httpx.AsyncClient() as http_client:
-        response = await http_client.post(url=url, json={"destination": address})
+        headers = {"Content-Type": "application/json"}
+        if use_case:
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:146]:
```suggestion
        if use_case is not None:
```

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 2 above)
Comment (justinr1234) [reply xrpl/asyncio/wallet/wallet_generation.py:34]: We suggested that it be moved to json instead of http header

**Comment by ckniffen** [comment]
Should there be a follow up PR which sets `use_case="integration_test"` or `use_case="unit_test"`?

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 2 above)
Comment (jonathanlei) [reply xrpl/asyncio/wallet/wallet_generation.py:34]: user_agent is specific to xrpl.js - use-case is pointing towards things like: workshop, int test, unit tests..

Comment (jonathanlei) [reply xrpl/asyncio/wallet/wallet_generation.py:34]: > Should there be a follow up PR which sets `use_case="integration_test"` or `use_case="unit_test"`?

I didn't see a unit test that call the fundWallet function

Comment (mvadari) [reply xrpl/asyncio/wallet/wallet_generation.py:34]: There aren't any

**Review by ckniffen — APPROVED** [review]

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 2 above)
Comment (ckniffen) [reply xrpl/asyncio/wallet/wallet_generation.py:34]: > > Should there be a follow up PR which sets `use_case="integration_test"` or `use_case="unit_test"`?
> 
> I didn't see a unit test that call the fundWallet function

No but generate_faucet_wallet is called in places.  That should allow for you pass use case.

Comment (mvadari) [reply xrpl/asyncio/wallet/wallet_generation.py:34]: > user_agent is specific to xrpl.js - use-case is pointing towards things like: workshop, int test, unit tests..

I'm a little confused as to what this means - `user_agent` is in the HTTP header and I think it's a pretty clear name

**Comment by ckniffen** [comment]
@jonathanlei pining about adding use_case to generate_faucet_wallet

**Comment by justinr1234** [comment]
@jonathanlei conflicts after latest master

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 4]
```diff
@@ -143,9 +148,14 @@ async def _check_wallet_balance(address: str, client: Client) -> int:
         raise
 
 
-async def _request_funding(url: str, address: str) -> None:
+async def _request_funding(
+    url: str, address: str, usage_context: Optional[str] = None
+) -> None:
     async with httpx.AsyncClient() as http_client:
-        response = await http_client.post(url=url, json={"destination": address})
+        json_body = {"destination": address, "userAgent": "xrpl-py"}
```
Comment (justinr1234) [inline xrpl/asyncio/wallet/wallet_generation.py:155]: I think user agent should also be an input parameter that defaults to xrpl-py but the user can change to be None if they want

**Review by justinr1234 — CHANGES_REQUESTED** [review]

**Review by ckniffen — APPROVED** [review]

**Review by justinr1234 — APPROVED** [review]

**Pull request merged**